### DISCUSSION TITLE
Idea: A new class name `group` for automatic radius inheritance (replacing `btn-group` and `input group`)

### DISCUSSION THREAD

**Discussion by saadeghi** [saadeghi/daisyui#1572] [Ideas / feature requests]
## Problem
Currently we have [`btn-group`](https://daisyui.com/components/button-group/) and [`input-group`](https://daisyui.com/components/input-group/). They are useful when we want to put multiple buttons in a row/column or multiple inputs and it applies border radius only for the first and last item and resets the border radius of middle ones.  
Like this:  
<img width="327" alt="1" src="https://user-images.githubusercontent.com/7342023/216420092-59075eb1-a9fd-440d-bc98-2fdc4cae463d.png">
and the syntax is like this:
```html
<div class="btn-group">
  <button class="btn">Button</button>
  <button class="btn">Button</button>
  <button class="btn">Button</button>
</div>
```

But there are problems:  
- ❌ Only the direct children of the group get the border radius. So if we put on of the children in a div or something, we won't get the correct result.
- ❌ `btn-group` and `input-group` are basically doing the same job so there's no reason for them to have different class names.
- ❌ The value of border radius is hardcoded. Which is not good.

## Solution

I want to add a new class name `group` to replace them with more flexibility and more customizability.
- ✅ The parent will have `group` class and the children should have `group-item` class. This will allow the `group-item` to get the style regardless of being a direct child or being in deeper DOM nodes.
- ✅ Having a `group` (instead of `btn-group` and `input-group`) is more semantic and it allows us to use any other element inside it. Not just inputs and buttons.
- ✅ the value of border radius will be inherited from the parent (`group`) so it will be customizable easily.

## Demo

This is my suggested syntax:
```html
<div class="group">
  <button class="btn group-item">Button</button>
  <button class="btn group-item">Button</button>
  <button class="btn group-item">Button</button>
</div>
```
And `group-item` doesn't need to be a direct child. So we can have other divs or wrapper elements (for example a daisyUI `tooltip`)
```html
<div class="group">
  <button class="btn group-item">Button</button>
  <button class="btn group-item">Button</button>
  <div>
    <div>
      <button class="btn group-item">Button</button>
    </div>
  </div>
</div>
```
Here is an early draft demo: https://play.tailwindcss.com/xaWNrndeVX

## Related issues

This will fix the following issues:
- #1467
- #1360
- #1254
- #1203
- #1200
- #894
- #877

## Backward compatibility

This new class name will replace `btn-group` and `input-group`. They will be removed from the docs but I still keep them in the package so if someone updates daisyUI, it won't break their existing UI.

## Give me feedbacks

Let me know what you think about this.
- Do the class names `group` and `group-item` sound good to you?
- Do you have any other suggestions?
- Is there anything I should consider?

**Comment by spencermize** [comment]
In concept, I think this makes sense. You might need to rethink the actual class name, though, as it might conflict with Tailwind's `group` class?

https://tailwindcss.com/docs/hover-focus-and-other-states#

**Reply by saadeghi** [reply]
You're right. I should find a better name. `group` is a good name but I don't want to have conflicts.

**Reply by saadeghi** [reply]
some ideas:
- `group-horizontal`/`group-vertical` and `group-item`
- `radius-group`/`radius-group-vertical` and `radius-item`
- `rounded-group`/`rounded-group-vertical` and `rounded-item`
- `radius-set`/`radius-set-vertical` and `radius-item`
- `radius-parent`/`radius-radius-vertical` and `radius-child`

**Reply by chazzox** [reply]
i think any `group-*` syntax would conflict with the way tailwind uses group.

in the past i've used `group-*` for nested groups with plugins as well as the fact that [`group-hover:`](https://tailwindcss.com/docs/hover-focus-and-other-states#styling-based-on-parent-state) is a thing. it would just feel wrong to use `group` as the prefix for this feature. still a cool idea tho!

**Reply by saadeghi** [reply]
what do you think about `radius-parent` and `radius-child`? I think it's the most descriptive option

**Reply by DannyDelott** [reply]
I like that the previous `btn-group` doesn't convey any specific styling semantics in the name, the radiuses are an implementation detail consumers don't need to know about.

How about `cluster` and `cluster-item`?

**Reply by saadeghi** [reply]
> How about `cluster` and `cluster-item`?

I like this one

**Comment by rathod-sahaab** [comment]
I like `radius-parent` mentioned in your reply to a comment. This is
- natural language
- can use `radius-ancestor` in future we also have siblings, cousins, grand etc. All the family tree at disposal

What are your thoughts about tailwind prettier plugin to format old code to new, that's just a wild Idea I don't know if that can be done but what if we could should we.

**Reply by saadeghi** [reply]
> What are your thoughts about tailwind prettier plugin to format old code to new, that's just a wild Idea I don't know if that can be done but what if we could should we.

It is possible but not with the existing [`prettier-plugin-tailwindcss`](https://www.npmjs.com/package/prettier-plugin-tailwindcss) because this plugin is only responsible for Tailwind CSS itself.  


<details>
<summary>
So we need a new plugin. But I don't think anyone would install a Perttier plugin just to replace 1 specific class name 😅
</summary>

![6fg3aicn6if41](https://user-images.githubusercontent.com/7342023/216457342-b5337463-3cb8-432a-bb8a-4901ec53c50a.png)

</details>

**Comment by HadiSDev** [comment]
Dont forget form-group
